akka-stream/src/main/scala/akka/stream/impl/Timers.scala

```diff
@@ -255,20 +255,27 @@ import akka.stream.stage._
             if (isClosed(in)) completeStage()
             else pull(in)
           } else {
-            val time = System.nanoTime
-            if (nextDeadline - time < 0) {
-              nextDeadline = time + timeout.toNanos
+            val now = System.nanoTime()
+            // Idle timeout triggered a while ago and we were just waiting for pull.
+            // In the case of now == deadline, the deadline has not passed strictly, but scheduling another thunk
+            // for that seems wasteful.
+            if (now - nextDeadline >= 0) {
+              nextDeadline = now + timeout.toNanos
               push(out, inject())
-            } else scheduleOnce(GraphStageLogicTimer, FiniteDuration(nextDeadline - time, TimeUnit.NANOSECONDS))
+            } else
+              scheduleOnce(GraphStageLogicTimer, FiniteDuration(nextDeadline - now, TimeUnit.NANOSECONDS))
           }
         }
 
         override protected def onTimer(timerKey: Any): Unit = {
-          val time = System.nanoTime
-          if ((nextDeadline - time < 0) && isAvailable(out)) {
-            push(out, inject())
-            nextDeadline = time + timeout.toNanos
-          }
+          val now = System.nanoTime()
+          // Timer is reliably cancelled if a regular element arrives first. Scheduler rather schedules too late
+          // than too early so the deadline must have passed at this time.
+          assert(
+            now - nextDeadline >= 0,
+            s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff ${now - nextDeadline}.")
```

Comment:
I guess that depends on the interpretation :) We wouldn't fail it because of an actual error but because our assumptions were wrong.

Before the stream was just stalled when it should be sending keep alives. From the viewpoint of the keep alives that's pretty bad. One level of interpretation higher you could probably say that's still not such a big problem, we wanted to keep it alive and it's been alive longer than expected... But even one level higher, we didn't want to keep it actually alive but wanted to send probes to kill the stream as soon as possible and the new behavior makes sense again...